### PR TITLE
dbt-materialize: support overriding generate_cluster_name

### LINKED ISSUE #27159: dbt-materialize: support `generate_cluster_name`
Outcome: support `generate_cluster_name` so users can prefix a cluster name per dbt target. E.g. so that when running in dev, the `dev_` cluster is used and in prod, the `prod_` cluster is used.

Context - [slack thread](https://materializeinc.slack.com/archives/C03REC5MEKU/p1715637139125189?thread_ts=1715613860.316119&cid=C03REC5MEKU):

Introduce a macro called `generate_cluster_name`, akin to `generate_schema_name`, that was invoked on every model’s cluster with the following default definition:
```
{% macro generate_cluster_name(custom_cluster_name, node) -%}

    {%- set default_cluster = target.cluster -%}
    {%- if custom_cluster_name is none -%}

        {{ default_cluster }}

    {%- else -%}

        {{ custom_cluster_name }}

    {%- endif -%}

{%- endmacro %}
```
We should automatically call `generate_cluster_name(model.cluster)` on each model’s cluster to determine the true cluster name to use.
The user can plug in their own version of the macro, like one that automatically prefixes each cluster with the target name:
```
{% macro generate_cluster_name(custom_cluster_name, node) -%}

    {%- set default_cluster = target.cluster -%}
    {%- if custom_cluster_name is none -%}

        {{ default_cluster }}

    {%- else -%}

        {{target.name}}_{{ custom_cluster_name }}

    {%- endif -%}

{%- endmacro %}
```

### PR DESCRIPTION
**Pull request by bobbyiliev**
### Motivation

Fixes #27159

This PR introduces a couple of things:
* A new `generate_deploy_cluster_name` macro responsible for appending the `_dbt_deploy` suffix for the blue/green deployments.
* A new `generate_final_cluster_name` adapter method.
* Users can define their own `generate_cluster_name` macro in their dbt project to override the cluster name generation logic, like one that automatically prefixes each cluster with the target name:

```jinja
{% macro generate_cluster_name(custom_cluster_name, node) -%}

    {%- set default_cluster = target.cluster -%}
    {%- if custom_cluster_name is none -%}

        {{ default_cluster }}

    {%- else -%}

        {{target.name}}_{{ custom_cluster_name }}

    {%- endif -%}

{%- endmacro %}
```

### CODE REVIEW

File: misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_await.sql [hunk 1]
```diff
@@ -38,7 +38,7 @@
 {% set clusters = target_config.get('clusters', []) %}
 
 {% for cluster in clusters %}
-    {% set deploy_cluster = cluster ~ "_dbt_deploy" %}
+    {% set deploy_cluster = generate_cluster_name_internal(cluster, force_deploy_suffix=True) %}
```
Comment (bobbyiliev) [inline misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_await.sql:41]: By using the `generate_cluster_name_internal` macro in the deploy macros instead of directly concatenating `cluster ~ "_dbt_deploy"`, we enable users to override the `generate_cluster_name` macro without disrupting the deploy workflow.

For example, if a user defines their own `generate_cluster_name` macro like this:

```jinja
{% macro materialize__generate_cluster_name(custom_cluster_name, node) -%}
    {%- set default_cluster = target.cluster -%}
    {%- if custom_cluster_name is none -%}
        {{ default_cluster }}
    {%- else -%}
        {{ target.name }}_{{ custom_cluster_name }}
    {%- endif -%}
{%- endmacro %}
```

The blue/green deployments will remain compatible with this custom logic. The deployment workflow will correctly generate names such as `dev_some_name_dbt_deploy`, where `dev` is the user-defined target name.

**Comment by benesch** [comment]
Nice! I think the idea of providing two macros is exactly right. Here's how I'd slice it, for maximum end user ergonomics.

First, I think we want to keep the macros as simple as possible. The first macro, `generate_cluster_name`, is exactly as you have it. It's used to customize base cluster names, before any blue/green suffixing happens. By default it just returns the cluster name unchanged:

```jinja2
{% macro materialize__generate_cluster_name(custom_cluster_name) -%}

    {{ custom_cluster_name }}

{%- endmacro %}
```

Then we add a new macro called `generate_deploy_cluster_name`. This is what adjusts a cluster name for a blue/green deployment. By default it appends `_dbt_deploy`, but users could override it to e.g. use a prefix instead, or use a different term entirely.

```
{% macro materialize__generate_deploy_cluster_name(custom_cluster_name) -%}

    {{ custom_cluster_name }}_dbt_deploy

{%- endmacro %}

```

Then, I think we can hide away the implementation of generating the final cluster name as a method on the `MaterializeAdapter` class. This will keep it more tucked away from end users, since it's not something they should ever have to think about. I _think_ it's something like this, though I haven't actually tried this code:

```patch
diff --git a/misc/dbt-materialize/dbt/adapters/materialize/impl.py b/misc/dbt-materialize/dbt/adapters/materialize/impl.py
index 53228150f0..830a9c44a0 100644
--- a/misc/dbt-materialize/dbt/adapters/materialize/impl.py
+++ b/misc/dbt-materialize/dbt/adapters/materialize/impl.py
@@ -291,3 +291,9 @@ class MaterializeAdapter(PostgresAdapter):
         self.connections.execute(f"drop view {view_name}")
 
         return columns
+
+    def generate_final_cluster_name(self, cluster_name: str, force_deploy_suffix: bool = False) -> str:
+        cluster_name = self.dispatch("generate_cluster_name", "materialize")(cluster_name)
+        if self.connections.profile.cli_vars.get("deploy", False) or force_deploy_suffix:
+            cluster_name = self.dispatch("generate_deploy_cluster_name", "materialize")(cluster_name)
+        return cluster_name

```

Then, in the macros to handle blue/green and materialization creation, you'd just call this adapter method (instead of `generate_cluster_name_internal`), like so:

```jinja2
{% set deploy_cluster = adapter.generate_final_cluster_name(cluster, force_deploy_suffix=True) %}
```

Let me know if that makes sense! I think it addresses @morsapaes's concerns about the preponderance of finicky macros while still giving the end user full control of the cluster naming.

**Comment by bobbyiliev** [comment]
This is an interesting idea, I like making the macros as simple as possible.

One thing that I'm not 100% sure about is the method on the adapter class, as it's not very clear in the dbt docs but `adapter.dispatch` seems to be only available as a Jinja function, so this will result in `type object 'MaterializeAdapter' has no attribute 'dispatch'`.

I will look further into this.

**Comment by bobbyiliev** [comment]
So far there doesn't seem to be a clear way of doing this without things getting too messy. Maybe I'm missing some context, I'll look into this further tomorrow morning.

**Comment by benesch** [comment]
> So far there doesn't seem to be a clear way of doing this without things getting too messy. Maybe I'm missing some context, I'll look into this further tomorrow morning.

It looks like there's an `execute_macro` method on `BaseAdapter` that should do what we need! Again untested, but does something like this work?

```diff
diff --git a/misc/dbt-materialize/dbt/adapters/materialize/impl.py b/misc/dbt-materialize/dbt/adapters/materialize/impl.py
index 53228150f0..830a9c44a0 100644
--- a/misc/dbt-materialize/dbt/adapters/materialize/impl.py
+++ b/misc/dbt-materialize/dbt/adapters/materialize/impl.py
@@ -291,3 +291,9 @@ class MaterializeAdapter(PostgresAdapter):
         self.connections.execute(f"drop view {view_name}")
 
         return columns
+
+    def generate_final_cluster_name(self, cluster_name: str, force_deploy_suffix: bool = False) -> str:
+        cluster_name = self.dispatch("generate_cluster_name", "materialize")(cluster_name)
+        if self.connections.profile.cli_vars.get("deploy", False) or force_deploy_suffix:
+            cluster_name = self.execute_macro("generate_deploy_cluster_name", kwargs={custom_cluster_name=cluster_name})
+        return cluster_name
```

**Comment by bobbyiliev** [comment]
> It looks like there's an execute_macro method on BaseAdapter that should do what we need! Again untested, but does something like this work?

Ah neat! Thanks! Last night I was going in a totally different direction.

Just tested this and it seems to be exactly what we need indeed! 

Just pushed those changes, and I've done some testing locally and it looks good, but totally open to feedback. Planning to do some more testing to make sure that I'm not missing some edge cases.

File: misc/dbt-materialize/dbt/adapters/materialize/impl.py [hunk 2]
```diff
@@ -291,3 +291,21 @@ def get_column_schema_from_query(self, sql: str) -> List[PostgresColumn]:
         self.connections.execute(f"drop view {view_name}")
 
         return columns
+
+    @available
+    def generate_final_cluster_name(
+        self, cluster_name: str, force_deploy_suffix: bool = False
+    ) -> str:
+        cluster_name = self.execute_macro(
+            "materialize__generate_cluster_name",
```
Comment (benesch) [inline misc/dbt-materialize/dbt/adapters/materialize/impl.py:300]: Does this need to include the `materialize__` prefix? I'm a little worried that will prevent people from overriding the macro. Does it not work to have the bare `generate_cluster_name`?

File: misc/dbt-materialize/dbt/adapters/materialize/impl.py [hunk 3]
```diff
@@ -291,3 +291,21 @@ def get_column_schema_from_query(self, sql: str) -> List[PostgresColumn]:
...
+            kwargs={"custom_cluster_name": cluster_name},
+        )
+        if (
+            self.connections.profile.cli_vars.get("deploy", False)
+            or force_deploy_suffix
+        ):
+            cluster_name = self.execute_macro(
+                "materialize__generate_deploy_cluster_name",
```
Comment (benesch) [inline misc/dbt-materialize/dbt/adapters/materialize/impl.py:308]: Ditto.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/generate_cluster_name.sql [hunk 4]
```diff
@@ -13,9 +13,18 @@
...
 
-    {%- set deploy_suffix = "_dbt_deploy" if var('deploy', False) else "" -%}
-    {{ default_cluster }}{{ deploy_suffix }}
+{% macro generate_cluster_name(custom_cluster_name) -%}
+    {{ return(adapter.dispatch('generate_cluster_name', 'materialize')(custom_cluster_name)) }}
+{%- endmacro %}
 
+{% macro materialize__generate_cluster_name(custom_cluster_name) -%}
+    {{ custom_cluster_name }}
 {%- endmacro %}
```
Comment (bobbyiliev) [inline misc/dbt-materialize/dbt/include/materialize/macros/utils/generate_cluster_name.sql:30]: As we just discussed with Nikhil, we came up with 2 suggestions here:
- Use 4 macros with the dispatch method, as how [dbt does it](https://github.com/dbt-labs/dbt-adapters/blob/a2292c8ec76e4c9f03869ad95817a2ad82dfb34b/dbt/include/global_project/macros/relations/drop.sql#L26-L34), and as currently is in this PR.
- An alternative option that we discussed is to use just two macros without using the `adapter.dispatch` method, eg:
    ```jinja
    {% macro generate_deploy_cluster_name(custom_cluster_name) -%}
        {{ custom_cluster_name }}_dbt_deploy
    {%- endmacro %}

    {% macro generate_cluster_name(custom_cluster_name) -%}
        {{ custom_cluster_name }}
    {%- endmacro %}
    ```
    
    As the macros are specific to our adapter, using just the two macros above without `adapter.dispatch`, will simplify things a bit. 

In both cases, people will be able to override this with:

```jinja
{% macro generate_cluster_name(custom_cluster_name, node) -%}

    {%- set default_cluster = target.cluster -%}
    {%- if custom_cluster_name is none -%}

        {{ default_cluster }}

    {%- else -%}

        {{target.name}}_{{ custom_cluster_name }}

    {%- endif -%}

{%- endmacro %}
```

Open to suggestions @morsapaes 🙌

**Review by benesch — APPROVED** [review]
This looks great to me! I still have that unresolved question about macro namespacing (why do we have to use `'materialize'` and not `'dbt_materialize'`), but I think the way we've split it up between macros and adapter methods is just right. Gives users a lot of flexibility but each of the macros is dead simple.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/generate_cluster_name.sql (hunk 4 above)
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/utils/generate_cluster_name.sql:30]: > An alternative option that we discussed is to use just two macros without using the adapter.dispatch method

This makes sense to me. We don't need to use the dispatch method, since that's only relevant for overriding core macros.

Comment (bobbyiliev) [reply misc/dbt-materialize/dbt/include/materialize/macros/utils/generate_cluster_name.sql]: Sounds good to me! I just updated the macros, so now we only have 2 macros left.

**Review by benesch — APPROVED** [review]
This turned out beautifully. Thanks so much, @bobbyiliev. No further questions. LGTM.

**Review by morsapaes — APPROVED** [review]
